enhancements/testing/openshift-tests-extension.md

```diff
@@ -0,0 +1,722 @@
+    # a given invocation of run-test if mode=instance).
+    "isolation": {
+      "mode": "exec",
+      "conflict":  [
+        "gpu",
+      ]
+    },
+      
+    # Tests that require significant resources can identify
+    # that fact so that the planning algorithm will not
+    # overdraw on allocatable pod resources for a set of
+    # tests run in parallel.
+    "memory": "1Gi",
+    
+    # If a duration can be roughly predicted, inclusion of this
+    # information may improve the test execution 
+    # planning algorithm performance.  
+    "duration": "2s",
+      
+    # Timeout information may also support efficient
+    # planning.
+    "timeout": "16s",
+  },
+
+  # Tests can be identified as "informing" or "production".
+  # Informing tests will not negatively impact default views
+  # in Component Readiness.
+  # However, policies may be established in the future that 
+  # cause tests that remain informing too long to be treated
+  # as if they are production.
+  "lifecycle": "informing",
+  
+}
+```
+
+##### Run-Test - Running Extension Tests
+The `run-test` verb is used to cause the extension binary to actually run 
+discovered tests. `run-test` accepts test environment arguments as discussed
+in the "list" verb section (in order to reduce fact discovery time needed
+for each test run) and one or more test names to invoke.
+```
+$ extension-binary run-test 
+  --platform      The hardware or cloud platform ("aws", "gcp", "metal", ...).
+  ...other context arguments...
+  --config        The component configuration profile to assume is active on the cluster.
+  --name | -n     Test name to invoke (-n can be specified multiple times).
+```
+
+`run-test` should run the specified tests in parallel. To prevent 
+test cases developing a dependence on one another, `run-test` must make no
+assumption about the number or composition of tests it is asked to run
+(other than they will be consistent with the test environment and isolation
+requirements defined by the "list" verb).
+
+`openshift-tests` will randomize tests into multiple different parallel
+executions of the extension binary in order to amortize initialization
+costs.
+
+Standard output from the invocation will be serialized into JSONL with
+each line associated with a single test run result.
+
+```python
+{ "name": "T1", "result": "success", }
+{ "name": "T1", "result": "success", }
+{ "name": "T1", "result": "success", }
+```
+
+A single test result will contain the following information, encoded on a single line:
+```python
+{
+    "name": "test name",
+    
+    # Duration of test run in milliseconds
+    "duration": "1012",
+    
+    # The outcome of the test; pass, fail, skip, timeout.
+    "result": "pass",
+    "output": "standard output of the test",
+    "error": "standard error of the test",
+    
+    # Human-readable messages to further explain 
+    # skips / timeouts / etc.
+    "messages": [
+        "example message"
+    ]
+}
+```

```

Comment:
Suggest some additional fields:

- startTime
- endTime

Reply:
Good call. Adding in the next push.

Reply:
Origin will decorate the results with the test spec in the aggregated output. So we will ultimately have this information in the DB.

Reply:
Should `ExtensionTestResult` just include the whole `ExtensionTestSpec`, so the results from invoking it from an extension binary produces the same output as if it was run from origin and aggregated?

Reply:
It will never be precisely the same. For example, only `origin` knows the full environment (e.g. context & full versions). There are pros/cons to including the spec in the result:
1. Con: The amount of output a human user of an extension binary must read through increases.
2. Pro: origin does not need to do the work of adding spec+result. 
3. Con: #2 is only true so long as the extension output is consistent with exactly what origin outputs in the aggregated format.  For example, if origin wants to move `component` to another place in the aggregated schema, it is free to do so today without having to remove anything from the extension result output from older versions of the extension. 
4. Pro: Consistency between origin and extension outputs.

The cons seemed to outweigh the pros, so I'm ever so slightly on the "bare minimum" result side.